receiver/receivercreator/discovery.go

```diff
@@ -0,0 +1,186 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package receivercreator // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/receivercreator"
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/go-viper/mapstructure/v2"
+	"go.uber.org/zap"
+	"gopkg.in/yaml.v3"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/extension/observer"
+)
+
+const (
+	// hints prefix
+	otelHints = "io.opentelemetry.discovery"
+
+	// hint suffix for metrics
+	otelMetricsHints = otelHints + ".metrics"
+
+	// hints definitions
+	discoveryEnabledHint = "enabled"
+	scraperHint          = "scraper"
+	configHint           = "config"
+)
+
+// hintsTemplatesBuilder creates configuration templates from provided hints.
+type hintsTemplatesBuilder interface {
+	createReceiverTemplateFromHints(env observer.EndpointEnv) (*receiverTemplate, error)
+}
+
+// k8sHintsBuilder creates configurations from hints provided as Pod's annotations.
+type k8sHintsBuilder struct {
+	logger          *zap.Logger
+	config          DiscoveryConfig
+	ignoreReceivers map[string]bool
+}
+
+var _ hintsTemplatesBuilder = (*k8sHintsBuilder)(nil)
+
+func createK8sHintsBuilder(config DiscoveryConfig, logger *zap.Logger) k8sHintsBuilder {
+	ignoreReceivers := make(map[string]bool, len(config.IgnoreReceivers))
+	for _, r := range config.IgnoreReceivers {
+		ignoreReceivers[r] = true
+	}
+	return k8sHintsBuilder{
+		logger:          logger,
+		config:          config,
+		ignoreReceivers: ignoreReceivers,
+	}
+}
+
+// createReceiverTemplateFromHints creates a receiver configuration based on the provided hints.
+// Hints are extracted from Pod's annotations.
+// Scraper configurations are only created for Port Endpoints.
+// TODO: Log receiver configurations are only created for Pod Container Endpoints.
+func (builder *k8sHintsBuilder) createReceiverTemplateFromHints(env observer.EndpointEnv) (*receiverTemplate, error) {
+	var endpointType string
```

Comment:
nit: this declaration seems redundant

Reply:
done: [046754a](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/35617/commits/046754a81ec3f31fa2a34a3725b796a4a5aa7804)